zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+#[scale_info(skip_type_params(T))]
+pub(crate) struct Node<T: Config> {
+    pub account: Option<T::AccountId>,
+    pub stake: BalanceOf<T>,
+    pub fees: BalanceOf<T>,
+    pub descendant_stake: BalanceOf<T>,
+    pub lazy_fees: BalanceOf<T>,
+}
+
+impl<T> Node<T>
+where
+    T: Config,
+{
+    /// Create a new node with `stake` belonging to `account`.
+    pub(crate) fn new(account: T::AccountId, stake: BalanceOf<T>) -> Node<T> {
+        Node {
+            account: Some(account),
+            stake,
+            fees: 0u8.into(),
+            descendant_stake: 0u8.into(),
+            lazy_fees: 0u8.into(),
+        }
+    }
+
+    /// Return the total stake of the node (the node's stake plus the sum of descendant's stakes).
+    pub(crate) fn total_stake(&self) -> Result<BalanceOf<T>, DispatchError> {
+        self.stake.checked_add_res(&self.descendant_stake)
+    }
+
+    /// Return `true` is the node is a leaf in the sense that none of its descendants hold any
+    /// stake. (Strictly speaking, it's not always a leaf!)
+    pub(crate) fn is_leaf(&self) -> bool {
+        self.descendant_stake == Zero::zero()
+    }
+}
+
+/// Execution path info for `join` calls.
+#[derive(Debug, PartialEq)]
+pub enum BenchmarkInfo {
+    /// The LP already owns a node in the tree.
+    InPlace,
+    /// The LP is reassigned an abandoned node.
+    Reassigned,
+    /// The LP is assigned a leaf of the tree.
+    Leaf,
+}
+
+impl<T, U> LiquiditySharesManager<T> for LiquidityTree<T, U>
+where
+    T: Config + frame_system::Config,
+    T::AccountId: PartialEq<T::AccountId>,
+    BalanceOf<T>: AtLeast32BitUnsigned + Copy + Zero,
+    U: Get<u32>,
+{
+    type JoinBenchmarkInfo = BenchmarkInfo;
+
+    fn join(
+        &mut self,
+        who: &T::AccountId,
+        stake: BalanceOf<T>,
+    ) -> Result<Self::JoinBenchmarkInfo, DispatchError> {
+        let index_maybe = self.account_to_index.get(who);
+        let (index, benchmark_info) = if let Some(&index) = index_maybe {
+            // Pile onto existing account.
+            self.propagate_fees_to_node(index)?;
+            let node = self.get_node_mut(index)?;
+            node.stake = node.stake.checked_add_res(&stake)?;
+            (index, BenchmarkInfo::InPlace)
+        } else {
+            // Push onto new node.
+            let (index, benchmark_info) = match self.peek_next_free_node_index()? {
+                NextNode::Abandoned(index) => {
+                    self.propagate_fees_to_node(index)?;
+                    let node = self.get_node_mut(index)?;
+                    node.account = Some(who.clone());
+                    node.stake = stake;
+                    node.fees = Zero::zero(); // Not necessary, but better safe than sorry.
+                    // Don't change `descendant_stake`; we're still maintaining it for abandoned
+                    // nodes.
+                    node.lazy_fees = Zero::zero();
+                    self.abandoned_nodes.pop();
```

Comment:
Don't you think my `debug_assert!` is useful 😭?

Reply:
Frankly, no. If the assert would ever trigger, that would be a disaster for the functionality of the tree. So, if anything, it should be an error.

But we have actual tests which verify that abandoned nodes work correctly. Adding checks which verify the correctness of the state at every corner of the code is an anti-pattern if you ask me. We just do this sometimes because we have no choice but to throw an error instead of panicking.

Anyways, the underlying problem here is that the whole peek-then-pop design is garbage. Fixed here: 8ea0f9caee85b3ac25a4ebe4ea1087bc3c1070cc

Reply:
> But we have actual tests which verify that abandoned nodes work correctly.


Ok, this is good enough.

Reply:
looks far easier to understand now https://github.com/zeitgeistpm/zeitgeist/commit/8ea0f9caee85b3ac25a4ebe4ea1087bc3c1070cc